compiler/src/iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertCommandBufferOps.cpp

```diff
@@ -88,6 +141,23 @@ class CommandBufferPushDescriptorSetOpConversion
       ConversionPatternRewriter &rewriter) const override {
     auto importType = importOp.getFunctionType();
 
+    Value zero;
+    auto getI32Zero = [&]() {
+      if (!zero) {
+        zero = rewriter.create<IREE::VM::ConstI32ZeroOp>(op.getLoc());
+      }
+      return zero;
+    };
+    Value null;
+    auto getNull = [&]() {
+      if (!null) {
+        null = rewriter.create<IREE::VM::ConstRefZeroOp>(
+            op.getLoc(),
+            IREE::VM::RefType::get(rewriter.getType<IREE::HAL::BufferType>()));
+      }
+      return null;
+    };
```

Comment:
Is this necessary? Wouldn't CSE/createOrFold handle uniqueing these 'zero' and 'null' ops? (and if they don't, should we make changes so they do?)

Reply:
CSE will take care of it but not until after the pass completes and CSE runs - every op we create, insert into the IR, run verification on, run canonicalization/folding on, and run CSE on has a big cost and this kind of stuff is pretty cheap - this here saves us 1000-10000 additional ops during lowering. I have IndexSet as a helper that most other dialects uses to do this in 1 line but in the VM we have to emit these ops directly so it can't be reused ;(

Reply:
(could be nice to generalize IndexSet into ConstantSet with template magic or something, though, so anything could be memoized)